### PR TITLE
New admob extension

### PR DESCRIPTION
**Pull request by Lizard-13**
Work in progress, now it cordova should install the admob plugin automatically when adding a platform (android, ios).

@4ian As you can see, the admob application ID (now needed obligatorily) is a project property, because it has to be unique per project and we have to access it when exporting.
The exporter adds the plugin configuration, the version is hardcoded because it's required and this version works so far (plugins world changes fast so maybe it will need an update in the future) . With this extra configuration cordova downloads the plugin automatically when a platform is added.
Let me know what do you think, because I know it's not excellent, but the obligatory application ID changes the game.

Now I'll write it as a pure JS extensions, as suggested. And I was thinking about not using an "AdMob" object at all, but just conditions and actions, everything with ads seems to be singletons or something like that, there's just one admob.rewardvideo element for example.

### CODE REVIEW

**Comment by 4ian** [comment]
That looks like a good start!

> As you can see, the admob application ID (now needed obligatorily) is a project property, because it has to be unique per project and we have to access it when exporting

It's fine to have it here.  I was also thinking about where is the best place to store this.
Ideally, we could maybe have something like `PlatformSpecificAssets`, a class that is able to store arbitrary data about a specific platform/extension. We could call it  `PlatformSpecificSettings`.
For now we only have adMobAppId, but we could image having other Cordova specific settings in the future.
In an even more ideal world, extensions would be able to extend the UI of the software, so that we would both have `PlatformSpecificSettings` to store new data like this id, and also functions to "plug" UI written in React inside the IDE at some places (project settings, scene settings mainly).

This being said, we can refactor this later. On the UI, we're still "hardcoding" the existence of this id because there is no UI plugin, so we might as well have it hardcoded in the project class.

> The exporter adds the plugin configuration, the version is hardcoded because it's required and this version works so far (plugins world changes fast so maybe it will need an update in the future) 

Hardcoding the plugin number is fine - and from experience the best thing to do to avoid breakages. :)

> Now I'll write it as a pure JS extensions, as suggested. And I was thinking about not using an "AdMob" object at all, but just conditions and actions, everything with ads seems to be singletons or something like that, there's just one admob.rewardvideo element for example.

Yeah it's also fine not to have an object (might be harder to discover but I think we could improve the dialog to find conditions/actions - or even display in other part of the software that such extensions exist.) 
This will better mirror the API and it was strange anyway to have to use an object while only one ad can be displayed.

**Comment by rikazukun** [comment]
Cool thanks, actually I am currently working on it right now also. With the new api changes the app ID on the manifest is indeed a requirement.
Cheers~

**Comment by bil-ash** [comment]
@Lizard-13 That's great. I tried with the plugin but could not get the rewarded video ads working(although banner and interstitial ads were working). Then I came across this page in the documentation-
https://ratson.github.io/cordova-plugin-admob-free/variable/index.html#static-variable-rewardvideo
where it was mentioned that for rewarded video ads **mediation plugin** is necessary,so I did not bother about trying with this anymore. But the apk whose link you have shared dosen't seem to have mediation plugin but shows video ads perfectly(at least on my phone). Great work

**Comment by Lizard-13** [comment]
Thanks!

@bil-ash About the mediation plugin, I need real tests in the future, if real ads don't work I hope fixing it is easy as adding the mediation plugin aside the admob one in the exporter. Or I'll have to do some extra research & implementation.

@4ian Yeah, haven't found the time to implement and test the banners and interstitials, the API is the same so far so I just have to copy the video code and rename, 1-2 hours of work I guess.

**Comment by 4ian** [comment]
@Lizard-13 Cool, let me know when you have a chance to implement the rest or if there is anything where you need help. :)
(Note that I will be less available until next Sunday as I'm traveling to make a talk about [GDevelop at React Conf](https://conf.reactjs.org/event.html?FlorianRival)).

**Comment by Lizard-13** [comment]
Ready, here is a working apk featuring banners, interstitials and reward videos:
[app-debug.zip](https://github.com/4ian/GDevelop/files/2508709/app-debug.zip)

@4ian Please review the event listeners part (and of course everything), not sure if I should/can remove the listeners at the end of the game and if it worths it :P

**Comment by oussama-HA** [comment]
@4ian @Lizard-13 so when we gonna see this in gdevelop ? Hope soon

**Comment by 4ian** [comment]
Hey, sorry for the delay I was traveling at a conference this week.

> @4ian Please review the event listeners part (and of course everything), not sure if I should/can remove the listeners at the end of the game and if it worths it :P

I think it's not worth removing them :) 

The rest looks good to me, I don't see anything that would be an issue.
If this looks ready, we can merge it and I'll remove the old extension for Admob. I'll double check with the build service, but as it's a standard Cordova plugin it should work as is.

**Comment by Lizard-13** [comment]
Yes, but lets wait I get some answer about the mediation plugin, so I add it right now and we don't publish a feature broken from the beginning :)

**Comment by 4ian** [comment]
Yep let's wait then :)

**Comment by zatsme** [comment]
This is great work and another feature for gd but I wonder if we need to go with the plus version that the author is maintaining as he's stated limited support for free and seems to be happy for us to integrate into gd?

**Comment by Lizard-13** [comment]
The mediation plugin is not needed at first, so we can merge it now to start the real tests :)

I can switch to the new admob-plugin when it gets more stable and I can make it work with the online build services.

File: Extensions/AdMob/JsExtension.js [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+ * More information on https://github.com/4ian/GDevelop/blob/master/newIDE/README-extensions.md
+ */
+module.exports = {
+  createExtension: function(t, gd) {
+    const extension = new gd.PlatformExtension();
+    extension.setExtensionInformation(
+      "AdMob",
+      t("AdMob"),
+      t(
+        "Allow the game to display AdMob banner, intersticial and reward video ads"
```
Comment (4ian) [inline Extensions/AdMob/JsExtension.js]: Typo here, it's "interstitial" ;)

**Comment by 4ian** [comment]
Super great!
Just found a typo. Will merge tomorrow and remove the old extension :) Will have to update the wiki too. 
Thanks for working on this! Will be super useful and appreciated by I think lots of user that would like to add ads in their games. Great addition to GDevelop.

File: Extensions/AdMob/JsExtension.js (hunk 1 above)
Comment (Lizard-13) [reply Extensions/AdMob/JsExtension.js]: Aha :)

**Comment by Lizard-13** [comment]
@4ian Here is the project I used to test the extension, the one that generated the apk:
[AdMobTest.zip](https://github.com/4ian/GDevelop/files/2531007/AdMobTest.zip)

**Comment by 4ian** [comment]
I've tried to build for Android but unfortunately got some errors 🤔Strange because I was thinking that the Cordova plugin would be self sufficient?

I've attached the logs (just added real line breaks for clarity)
[build-log.txt](https://github.com/4ian/GDevelop/files/2532182/build-log.txt)

I've seen that it could be related to `play-services-ads` (https://stackoverflow.com/questions/20403136/error-package-com-google-android-gms-ads-does-not-exist) but not sure why it was working for you?

**Comment by 4ian** [comment]
Might be related to Cordova version, could you tell me what was yours? I might have to upgrade the one from the build service :)

**Comment by Lizard-13** [comment]
I hope so, I've cordova-cli 8.1.2 and cordova-android 7.1.1 (the latest ones)

>Plugin doesn't support this project's cordova-android version. cordova-android: 6.1.2, failed version requirement
Skipping 'cordova-admob-sdk' for android

If the compiler error is because it can't install the admob plugin, it's a shame it can't just ignore all the admob realted imports, because these requirements are added by the plugin itself :/

**Comment by 4ian** [comment]
> Plugin doesn't support this project's cordova-android version. cordova-android: 6.1.2, failed version requirement

Oh that's the reason, I missed this line! Will upgrade Cordova :)

**Comment by oussama-HA** [comment]
@Lizard-13 @4ian Good job guys. Please can you update the wiki or do some tutorials on this; I have downloaded the project but i am having a hard time to figure out to implement this

**Comment by 4ian** [comment]
@Lizard-13 It's working perfectly by setting the Cordova version to 8. 👍

I've added your test game to the examples, it's super useful! Great job again on this, this will be great plus to be able to add AdMob ads out of the box to games!

@oussama-HA You're running the development version? If not you have to wait for the next version (release next week I think). I'll update the wiki at this time.

**Comment by Lizard-13** [comment]
@oussama-HA Note that there is no need to do a specific "build admob project with cordova" article because there is no extra work to do, right now you just have to export the project with GD locally and add&build android/ios platforms with cordova, the plugin is installed automatically :raised_hands:

**Comment by 4ian** [comment]
FYI, I plan to release a new version in the next days with this :)

**Pull request merged**